dot/sync/fullsync.go

```diff
@@ -0,0 +1,584 @@
+
+	repChanges = make([]Change, 0)
+	peersToBlock = make([]peer.ID, 0)
+	validRes = make([]RequestResponseData, 0, len(results))
+
+resultLoop:
+	for _, result := range results {
+		request := result.request.(*messages.BlockRequestMessage)
+
+		if !result.completed {
+			continue
+		}
+
+		response := result.response.(*messages.BlockResponseMessage)
+		if request.Direction == messages.Descending {
+			// reverse blocks before pre-validating and placing in ready queue
+			slices.Reverse(response.BlockData)
+		}
+
+		err := validateResponseFields(request, response.BlockData)
+		if err != nil {
+			logger.Warnf("validating fields: %s", err)
+			// TODO: check the reputation change for nil body in response
+			// and nil justification in response
+			if errors.Is(err, errNilHeaderInResponse) {
+				repChanges = append(repChanges, Change{
+					who: result.who,
+					rep: peerset.ReputationChange{
+						Value:  peerset.IncompleteHeaderValue,
+						Reason: peerset.IncompleteHeaderReason,
+					},
+				})
+			}
+
+			continue
+		}
+
+		// only check if the responses forms a chain if the response contains the headers
+		// of each block, othewise the response might only have the body/justification for
+		// a block
+		if request.RequestField(messages.RequestedDataHeader) && !isResponseAChain(response.BlockData) {
+			logger.Warnf("response from %s is not a chain", result.who)
+			repChanges = append(repChanges, Change{
+				who: result.who,
+				rep: peerset.ReputationChange{
+					Value:  peerset.IncompleteHeaderValue,
+					Reason: peerset.IncompleteHeaderReason,
+				},
+			})
+			continue
+		}
+
+		for _, block := range response.BlockData {
+			if slices.Contains(badBlocks, block.Hash.String()) {
+				logger.Warnf("%s sent a known bad block: #%d (%s)",
+					result.who, block.Number(), block.Hash.String())
+
+				peersToBlock = append(peersToBlock, result.who)
+				repChanges = append(repChanges, Change{
+					who: result.who,
+					rep: peerset.ReputationChange{
+						Value:  peerset.BadBlockAnnouncementValue,
+						Reason: peerset.BadBlockAnnouncementReason,
+					},
+				})
+
+				continue resultLoop
+			}
+		}
+
+		validRes = append(validRes, RequestResponseData{
+			req:          request,
+			responseData: response.BlockData,
+		})
+	}
+
+	return repChanges, peersToBlock, validRes
+}
+
+// sortFragmentsOfChain will organise the fragments
+// in a way we can import the older blocks first also guaranting that
+// forks can be imported by organising them to be after the main chain
+//
+// e.g: consider the following fragment of chains
+// [ {17} {1, 2, 3, 4, 5} {6, 7, 8, 9, 10} {8} {11, 12, 13, 14, 15, 16} ]
+//
+// note that we have fragments with single blocks, fragments with fork (in case of 8)
+// after sorting these fragments we end up with:
+// [ {1, 2, 3, 4, 5}  {6, 7, 8, 9, 10}  {8}  {11, 12, 13, 14, 15, 16}  {17} ]
+func sortFragmentsOfChain(fragments [][]*types.BlockData) [][]*types.BlockData {
```

Comment:
given that this modifies `fragments`, I think it makes sense to omit the return param.